src/token/src/canister/is20_notify.rs

```diff
@@ -3,85 +3,63 @@
 use crate::canister::TokenCanister;
 use crate::types::{TxError, TxReceipt, TxRecord};
 use candid::{CandidType, Deserialize, Nat, Principal};
-use ic_canister::virtual_canister_call;
+use ic_canister::virtual_canister_call_oneway;
 use ic_cdk::api::call::CallResult;
 
-pub(crate) async fn notify(canister: &TokenCanister, transaction_id: Nat) -> TxReceipt {
-    let tx = {
-        let mut state = canister.state.borrow_mut();
-        let tx = state
-            .ledger
-            .get(&transaction_id)
-            .ok_or(TxError::TransactionDoesNotExist)?;
-
-        // We remove the notification here to prevent a concurrent call from being able to send the
-        // notification again (while this call is await'ing). If the notification fails, we add the id
-        // backed into the pending notifications list.
-        if !state.notifications.remove(&transaction_id) {
-            return Err(TxError::AlreadyNotified);
-        }
-
-        tx
-    };
-
-    match send_notification(&tx).await {
-        Ok(()) => Ok(tx.index),
-        Err((_, _)) => {
-            canister
-                .state
-                .borrow_mut()
-                .notifications
-                .insert(transaction_id);
-            Err(TxError::NotificationFailed)
-        }
-    }
-}
-
-pub(crate) async fn transfer_and_notify(
+pub(crate) fn approve_and_notify(
     canister: &TokenCanister,
-    to: Principal,
-    amount: Nat,
-    fee_limit: Option<Nat>,
+    spender: Principal,
+    value: Nat,
 ) -> TxReceipt {
-    let id = canister.transfer(to, amount, fee_limit)?;
-    notify(canister, id).await
+    let transaction_id = canister.approve(spender, value)?;
+    let state = canister.state.borrow_mut();
+    let tx = state
+        .ledger
+        .get(&transaction_id)
+        .ok_or(TxError::TransactionDoesNotExist)?;
+    match send_notification(&tx) {
+        Ok(()) => Ok(tx.index),
+        Err((_, _)) => Err(TxError::NotificationFailed),
+    }
```

Comment:
We also want the error if the notification failed, this tells the user to retry the action. Perhaps  the solution to this is to separate the approve and notify steps, I think actually this would be a better and more general solution, as it would enable a user to notify about any transaction, be it a burn, transfer or mint.